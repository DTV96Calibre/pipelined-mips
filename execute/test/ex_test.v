`include "../execute_stage.v"

/**
 * Tests the functionality of the execute stage of our pipelined processor.
 */
module ex_test();

reg clk;
reg FlushE;
reg RegWriteD;
reg MemtoRegD;
reg MemWriteD;
reg [3:0] ALUControlD;
reg ALUSrcD;
reg RegDstD;
reg [31:0] RD1;
reg [31:0] RD2;
reg [4:0] RsD;
reg [4:0] RtD;
reg [4:0] RdD;
reg [31:0] SignImmD;
reg [31:0] ResultW;
reg [31:0] ALUOutM;
reg [1:0] ForwardAE;
reg [1:0] ForwardBE;

wire RegWriteE;
wire MemtoRegE;
wire MemWriteE;
wire [3:0] ALUControlE;
wire ALUSrcE;
wire RegDstE;
wire [31:0] RD1E;
wire [31:0] RD2E;
wire [4:0] RsE;
wire [4:0] RtE;
wire [4:0] RdE;
wire [31:0] SignImmE;
wire [4:0] WriteRegE;
wire [31:0] WriteDataE;
wire [31:0] ALUOutE;

initial begin
    // Init 'inputs'
    FlushE = 0;
    RegWriteD = 0;
    MemtoRegD = 0;
    MemWriteD = 0;
    ALUControlD = 4'b0010;
    ALUSrcD = 0;
    RegDstD = 0;
    RD1 = 32'b1;
    RD2 = 32'b11;
    RsD = 5'b00100;
    RtD = 5'b00101;
    RdD = 5'b00110;
    SignImmD = 32'b0;
    ResultW = 32'b111;
    ALUOutM = 32'b10;
    ForwardAE = 0;
    ForwardBE = 0;
    clk = 0;

    #1 clk = 1;
    #1 RegWriteD = 1;
    #1 RD1 = 32'b1110;
end

execute_stage EX_stage(clk, FlushE, RegWriteD, MemtoRegD, MemWriteD, ALUControlD,
    ALUSrcD, RegDstD, RD1, RD2, RsD, RtD, RdD, SignImmD,
    RegWriteE, MemtoRegE, MemWriteE, ALUControlE, ALUSrcE, RegDstE,
    RD1E, RD2E, RsE, RtE, RdE, SignImmE,
    ResultW, ALUOutM, ForwardAE, ForwardBE, 
    WriteRegE, WriteDataE, ALUOutE);

always @(*)
begin
    $display("display ALUOutE: %h", ALUOutE);
end

endmodule

