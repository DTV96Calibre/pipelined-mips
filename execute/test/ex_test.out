#! /usr/remote/bin/vvp
:ivl_version "10.1 (stable)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision + 0;
:vpi_module "system";
:vpi_module "vhdl_sys";
:vpi_module "v2005_math";
:vpi_module "va_math";
S_0x1f47760 .scope module, "alu" "alu" 2 9;
 .timescale 0 0;
    .port_info 0 /INPUT 32 "lvalue"
    .port_info 1 /INPUT 32 "rvalue"
    .port_info 2 /INPUT 4 "aluOP"
    .port_info 3 /OUTPUT 32 "result"
o0x7ff8f2acc018 .functor BUFZ 4, C4<zzzz>; HiZ drive
v0x1f47950_0 .net "aluOP", 3 0, o0x7ff8f2acc018;  0 drivers
o0x7ff8f2acc048 .functor BUFZ 32, C4<zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz>; HiZ drive
v0x1f59fe0_0 .net "lvalue", 31 0, o0x7ff8f2acc048;  0 drivers
v0x1f5a0c0_0 .var "result", 31 0;
o0x7ff8f2acc0a8 .functor BUFZ 32, C4<zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz>; HiZ drive
v0x1f5a1b0_0 .net "rvalue", 31 0, o0x7ff8f2acc0a8;  0 drivers
v0x1f5a290_0 .var "truevall", 31 0;
v0x1f5a3c0_0 .var "truevalr", 31 0;
E_0x1f48790 .event edge, v0x1f5a1b0_0, v0x1f59fe0_0;
    .scope S_0x1f47760;
T_0 ;
    %wait E_0x1f48790;
    %load/vec4 v0x1f59fe0_0;
    %cmpi/e 4294967295, 4294967295, 32;
    %jmp/0xz  T_0.0, 4;
    %pushi/vec4 0, 0, 32;
    %store/vec4 v0x1f5a290_0, 0, 32;
    %jmp T_0.1;
T_0.0 ;
    %load/vec4 v0x1f59fe0_0;
    %store/vec4 v0x1f5a290_0, 0, 32;
T_0.1 ;
    %load/vec4 v0x1f5a1b0_0;
    %cmpi/e 4294967295, 4294967295, 32;
    %jmp/0xz  T_0.2, 4;
    %pushi/vec4 0, 0, 32;
    %store/vec4 v0x1f5a3c0_0, 0, 32;
    %jmp T_0.3;
T_0.2 ;
    %load/vec4 v0x1f5a1b0_0;
    %store/vec4 v0x1f5a3c0_0, 0, 32;
T_0.3 ;
    %load/vec4 v0x1f47950_0;
    %dup/vec4;
    %pushi/vec4 2, 0, 4;
    %cmp/u;
    %jmp/1 T_0.4, 6;
    %dup/vec4;
    %pushi/vec4 6, 0, 4;
    %cmp/u;
    %jmp/1 T_0.5, 6;
    %dup/vec4;
    %pushi/vec4 1, 0, 4;
    %cmp/u;
    %jmp/1 T_0.6, 6;
    %dup/vec4;
    %pushi/vec4 0, 0, 4;
    %cmp/u;
    %jmp/1 T_0.7, 6;
    %dup/vec4;
    %pushi/vec4 7, 0, 4;
    %cmp/u;
    %jmp/1 T_0.8, 6;
    %dup/vec4;
    %pushi/vec4 15, 15, 4;
    %cmp/u;
    %jmp/1 T_0.9, 6;
    %jmp T_0.10;
T_0.4 ;
    %load/vec4 v0x1f5a290_0;
    %load/vec4 v0x1f5a3c0_0;
    %add;
    %store/vec4 v0x1f5a0c0_0, 0, 32;
    %jmp T_0.10;
T_0.5 ;
    %load/vec4 v0x1f5a290_0;
    %load/vec4 v0x1f5a3c0_0;
    %sub;
    %store/vec4 v0x1f5a0c0_0, 0, 32;
    %jmp T_0.10;
T_0.6 ;
    %load/vec4 v0x1f5a290_0;
    %load/vec4 v0x1f5a3c0_0;
    %or;
    %store/vec4 v0x1f5a0c0_0, 0, 32;
    %jmp T_0.10;
T_0.7 ;
    %load/vec4 v0x1f5a290_0;
    %load/vec4 v0x1f5a3c0_0;
    %and;
    %store/vec4 v0x1f5a0c0_0, 0, 32;
    %jmp T_0.10;
T_0.8 ;
    %load/vec4 v0x1f5a290_0;
    %load/vec4 v0x1f5a3c0_0;
    %cmp/u;
    %flag_get/vec4 5;
    %pad/u 32;
    %store/vec4 v0x1f5a0c0_0, 0, 32;
    %jmp T_0.10;
T_0.9 ;
    %pushi/vec4 4294967295, 4294967295, 32;
    %store/vec4 v0x1f5a0c0_0, 0, 32;
    %jmp T_0.10;
T_0.10 ;
    %pop/vec4 1;
    %jmp T_0;
    .thread T_0, $push;
# The file index is used to find the file name in the following table.
:file_names 3;
    "N/A";
    "<interactive>";
    "..//alu.v";
